scheme/src/main/java/cz/o2/proxima/scheme/confluent/SchemaRegistryValueSerializer.java

```diff
@@ -0,0 +1,207 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Slf4j
+class SchemaRegistryValueSerializer<M extends GenericContainer>
+    implements ValueSerializer<M> {
+
+  static final byte MAGIC_BYTE = 0x0;
+  static final int SCHEMA_ID_SIZE = 4;
+  private final URI schemaRegistryUri;
+  private transient SchemaRegistryClient schemaRegistry = null;
+  private Class<M> clazz = null;
+  private Integer schemaId = null;
+
+  private transient Map<Integer, AvroSerializer> serializersCache = null;
+
+
+  SchemaRegistryValueSerializer(URI scheme) throws URISyntaxException {
+    this.schemaRegistryUri = new URI(scheme.getSchemeSpecificPart());
+  }
+
+  @Override
+  public Optional<M> deserialize(byte[] input) {
+    if (input.length == 0) {
+      /**
+       * This is a little bit weird but proxima try to deserialize empty bytes
+       * in serializer validation {@link ValueSerializer#isValid(byte[])}
+       */
+      return Optional.of(getDefault());
+    }
+
+    try {
+      return deserializeValue(input);
+    } catch (Exception e) {
+      throw new SerializationException("Unable to deserialize value.", e);
+    }
+  }
+
+  @Override
+  public byte[] serialize(M value) {
+    try {
+      return serializeValue(value, getSchemaId());
+    } catch (IOException e) {
+      throw new SerializationException("Unable to serialize data.", e);
+    }
+  }
+
+  @Override
+  public M getDefault() {
+    try {
+      return getClassType().newInstance();
+    } catch (InstantiationException | IllegalAccessException e) {
+      throw new SerializationException("Unable to create default instance.", e);
+    }
+  }
+
+  @Override
+  public Class<M> getClassType() {
+    return getAvroClass();
+  }
+
+  @SuppressWarnings("unchecked")
+  private <M extends GenericContainer> byte[]
+      serializeValue(M value, int schemaId) throws IOException {
+
+    final ByteArrayOutputStream out = new ByteArrayOutputStream();
+    out.write(MAGIC_BYTE);
+    out.write(ByteBuffer.allocate(SCHEMA_ID_SIZE).putInt(schemaId).array());
+    AvroSerializer<M> avroSerializer = (AvroSerializer<M>) getAvroserializersCache()
+        .computeIfAbsent(schemaId, this::createSerializer);
+    avroSerializer.serialize(value, out);
+    out.flush();
+    return out.toByteArray();
+  }
+
+  @SuppressWarnings("unchecked")
+  private <M extends GenericContainer> Optional<M>
+      deserializeValue(byte[] bytes) throws IOException {
+
+    ByteBuffer buffer = getByteBuffer(bytes);
+    int dataSchemaId = buffer.getInt();
+    int len = buffer.limit() - 1 - SCHEMA_ID_SIZE;
+    int start = buffer.position() + buffer.arrayOffset();
+    AvroSerializer<M> avroSerializer = (AvroSerializer<M>) getAvroserializersCache()
+        .computeIfAbsent(dataSchemaId, this::createSerializer);
+    return Optional.of(avroSerializer.deserialize(buffer, start, len));
```

Comment:
The deserialization should not throw exceptions on malformed inputs. It should rather return `Optional.empty()`. That ensures graceful handling of malformed inputs (you most likely don't want to stop streaming pipeline just because one invalid input record).

Reply:
Fixed